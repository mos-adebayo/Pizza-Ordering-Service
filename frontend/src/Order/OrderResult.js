import React from 'react';
import PropTypes from 'prop-types';
import { Table, Badge } from "react-bootstrap";
import { appHelpers } from "../util/helper"
import Pagination  from "../util/pagination";

const OrderResult = ({ orders, activePage, pageLimit, totalOrders, handlePageClick }) => {
   const pageIndex = pageLimit * (activePage - 1);
    return (
        <>
            <Table striped bordered hover>
            <thead>
            <tr>
                <th>#</th>
                <th>Customer</th>
                <th>Pizza Type </th>
                <th>Pizza Size</th>
                <th>Quantity</th>
                <th>Status</th>
                <th>Address</th>
                <th>Date Ordered</th>
            </tr>
            </thead>
            <tbody>
            {
                orders.map((order, index) => <tr key={index}>
                    <td>{ (pageIndex) + (index + 1) }</td>
                    <td className={'text-capitalize'}>{`${order.customerName}`}</td>
                    <td>{`${order.pizzaType.name}`}</td>
                    <td>{`${order.pizzaSize.name}`}</td>
                    <td>{`${order.quantity}`}</td>
                    <td>
                        {
                            (order.status.name === "Pending") ? (
                                <Badge variant="warning">Pending</Badge>
                            ): (
                                <Badge variant="primary">{order.status.name}</Badge>
                            )
                        }
                    </td>
                    <td>{`${order.address}`}</td>
                    <td>{`${appHelpers.formatDate(order.createdAt)}`}</td>
                </tr>)
            }
            </tbody>
        </Table>
            
            <Pagination 
            activePage={activePage}
            totalItems={totalOrders} 
            handlePageClick={handlePageClick}
            pageLimit={pageLimit}
        />
        </>
    );
};

OrderResult.propTypes = {
    orders: PropTypes.array.isRequired
};

export default OrderResult;