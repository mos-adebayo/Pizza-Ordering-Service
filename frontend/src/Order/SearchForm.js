import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Field, ErrorMessage } from 'formik';
import { Form, Button, Row, Col, Spinner } from  "react-bootstrap";
import * as Yup from "yup";

const validationSchema = Yup.object().shape({
    customerName: Yup.string().required('Please enter customer name'),
    address: Yup.string().required('Please enter your address'),
    quantity: Yup.number()
        .required("Please enter quantity")
        .positive("Please enter valid quantity")
        .integer("Please enter valid quantity"),
    pizzaSizeId: Yup.string().required('Please select size'),
    pizzaTypeId: Yup.string().required('Please select type'),
});

const initialValues = { customerName: "", quantity: "", pizzaSizeId: "", pizzaTypeId: "", address: "" };

const SearchForm = ({ pizzaSizes, pizzaTypes, formHandler, requesting }) => {
    return (
        <Formik
            initialValues={initialValues}
            validationSchema={validationSchema}
            onSubmit={(values, ...props ) => {
                formHandler(values, props[0].resetForm);
            }}
        >
            {({ handleSubmit }) => (
                <form onSubmit={handleSubmit}>
                    <Row>
                        <Col>
                            <Form.Group>
                                <Form.Label> Customer </Form.Label>
                                <Field type="text" name="customerName" placeholder={"Enter customer name"} className={'form-control'} />
                                <ErrorMessage name="customerName" component="div" className={"text-danger"} />
                            </Form.Group>
                        </Col>
                        <Col>
                            <Form.Group>
                                <Form.Label> Quantity </Form.Label>
                                <Field type="text" name="quantity" placeholder="Enter quantity" className={'form-control'} />
                                <ErrorMessage name="quantity" component="div" className={"text-danger"}/>
                            </Form.Group>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <Form.Group>
                                <Form.Label> Pizza Size </Form.Label>
                                <Field component="select" name="pizzaSizeId" className={'form-control'} >
                                    <option value="-1">Select size</option>
                                    {
                                        pizzaSizes.map((item, index) => <option value={item.id} key={index}>
                                            {`${item.name}` }
                                        </option>)
                                    }
                                </Field>
                                <ErrorMessage name="pizzaSizeId" component="div"  className={"text-danger"}/>
                            </Form.Group>
                        </Col>
                        <Col>
                            <Form.Group>
                                <Form.Label> Pizza Type </Form.Label>
                                <Field component="select" name="pizzaTypeId" className={'form-control'} >
                                    <option value="-1">Select type</option>
                                    {
                                        pizzaTypes.map((item, index) => <option value={item.id} key={index}>
                                            {`${item.name}` }
                                        </option>)
                                    }
                                </Field>
                                <ErrorMessage name="pizzaTypeId" component="div" className={"text-danger"} />
                            </Form.Group>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <Form.Group>
                                <Form.Label> Address </Form.Label>
                                <Field type="text" name="address" placeholder={"Enter address"} className={'form-control'} />
                                <ErrorMessage name="address" component="div" className={"text-danger"} />
                            </Form.Group>
                        </Col>
                    </Row>
                    <Row>
                        <Col>
                            <Form.Group>
                                {
                                    requesting ? (
                                        <Button disabled>
                                            <Spinner animation="grow" />
                                        </Button>
                                    ) : (
                                        <Button type="submit" className={'submit-btn'}>
                                            Make Order
                                        </Button>
                                    )
                                }
                            </Form.Group>
                        </Col>
                    </Row>
                </form>
            )}
        </Formik>
    );
};

SearchForm.propTypes = {
    pizzaSizes: PropTypes.array.isRequired,
    pizzaTypes: PropTypes.array.isRequired,
    formHandler: PropTypes.func.isRequired
};

export default SearchForm;