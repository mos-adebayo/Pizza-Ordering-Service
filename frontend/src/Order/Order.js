import React, { useEffect, useState } from 'react';
import logo from "../logo.svg";
import SearchForm from "./SearchForm";
import OrderList from "./OrderResult";
import { Alert } from "react-bootstrap";
import { getPizzaSizes, searchOrders, createOrder, getPizzaTypes } from "../util/common.service"
import {appConstants} from "../util/constant";

const pageLimit = 5;
let searchParams = { limit: pageLimit, page: 1 };

const Order = () => {
    const [errorMessage, setErrorMessage] = useState(false);
    const [requesting, setRequesting] = useState(false);

    const [pizzaSizes, setPizzaSizes] = useState([]);
    const [pizzaTypes, setPizzaTypes] = useState([]);
    const [orders, setOrders] = useState([]);

    // Pagination
    const [totalOrders, setTotalOrders] = useState(0);
    const [activePage, setActivePage] = useState(1);

    const setResponse = (response, callback) => {
        if(response.status === appConstants.REQUEST_SUCCESS){
            callback(response.response.data)
        }else {
            setErrorMessage(response.error.error)
        }
    };
    const setOrdersResponse = (response) => {
        if(response.status === appConstants.REQUEST_SUCCESS){
            setTotalOrders(response.response.data.totalRecords);
            setOrders(response.response.data.rows)
        }else {
            setErrorMessage(response.error.error)
        }
    };
    
    useEffect( () => {
        setRequesting(true);
        Promise.all([getPizzaSizes(), getPizzaTypes(), searchOrders(searchParams)])
            .then(([pizzaSizesResponse, pizzaTypesResponse, ordersResponse]) => {
                setResponse(pizzaSizesResponse, setPizzaSizes);
                setResponse(pizzaTypesResponse, setPizzaTypes);
                setOrdersResponse(ordersResponse);
                        setRequesting(false);
            }).catch(err => {
               setErrorMessage("Ensure to start the API service");
               setRequesting(false);
        })
    }, []);
    
    const fetchPageOrders = async (pageNumber) => {
        setRequesting(false);
        try{
            const orders = await searchOrders({...searchParams, page: pageNumber});
            setOrdersResponse(orders);
            setActivePage(pageNumber);
            setRequesting(false)
        }catch (error) {
            setErrorMessage("Unable to get records for page");
            setRequesting(false)
        }
    };
    
    const handleSubmit = async (values, resetCallBack) => {
        setRequesting(true);
        try{
            await createOrder(values);
            const orders = await searchOrders(searchParams);
            setOrdersResponse(orders);
            resetCallBack();
            setRequesting(false)
        } catch (e) {
            setRequesting(false);
        }

    };
    
    return (
        <div className="App">
            <div className="App-body">
                <header className={'text-center'}>
                    <img src={logo} className="App-logo" alt="logo" />
                </header>

                {/*    Error Message */}
                {
                    errorMessage &&
                    <Alert variant={"danger"}>
                        {errorMessage}
                    </Alert>
                }
                <SearchForm 
                    pizzaSizes={pizzaSizes}
                    pizzaTypes={pizzaTypes}    
                    formHandler={handleSubmit}
                    requesting={requesting}
                />

                
                <OrderList
                    orders={orders}
                    totalOrders={totalOrders}
                    activePage={activePage}
                    pageLimit={pageLimit}
                    handlePageClick={fetchPageOrders}
                />
            </div>
        </div>
    );
};

export default Order;