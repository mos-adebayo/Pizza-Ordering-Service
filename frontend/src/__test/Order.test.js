import React from "react";
import ReactDOM from "react-dom";
import { configure } from "enzyme";
import renderer from "react-test-renderer";
import Adapter from 'enzyme-adapter-react-16';
import Order  from "../Order/Order";

//Enzyme Configurations
configure({ adapter: new Adapter() });

test('Search Form renders without crashing', () => {

    const div = document.createElement('div');
    ReactDOM.render(
         <Order/>,
        div);
    ReactDOM.unmountComponentAtNode(div);
});

test('Search Form components snapshot', () => {

    const component = renderer.create(
        <Order />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});