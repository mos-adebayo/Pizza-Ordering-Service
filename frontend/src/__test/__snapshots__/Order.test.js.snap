// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form components snapshot 1`] = `
<div
  className="App"
>
  <div
    className="App-body"
  >
    <header
      className="text-center"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
    </header>
    <form
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
               Customer 
            </label>
            <input
              className="form-control"
              name="customerName"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
               Quantity 
            </label>
            <input
              className="form-control"
              name="quantity"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
               Pizza Size 
            </label>
            <select
              className="form-control"
              name="pizzaSizeId"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            >
              <option
                value="-1"
              >
                Select size
              </option>
            </select>
          </div>
        </div>
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
               Pizza Type 
            </label>
            <select
              className="form-control"
              name="pizzaTypeId"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            >
              <option
                value="-1"
              >
                Select type
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
               Address 
            </label>
            <input
              className="form-control"
              name="address"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <button
              className="submit-btn btn btn-primary"
              disabled={false}
              type="submit"
            >
              Make Order
            </button>
          </div>
        </div>
      </div>
    </form>
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            #
          </th>
          <th>
            Customer
          </th>
          <th>
            Pizza Type 
          </th>
          <th>
            Pizza Size
          </th>
          <th>
            Quantity
          </th>
          <th>
            Status
          </th>
          <th>
            Address
          </th>
          <th>
            Date Ordered
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <ul
      className="pagination"
    />
  </div>
</div>
`;
